core/trino-main/src/test/java/io/trino/execution/scheduler/TestFullNodeCapableNodeAllocator.java

```diff
@@ -0,0 +1,698 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.scheduler;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import io.airlift.units.DataSize;
+import io.trino.Session;
+import io.trino.client.NodeVersion;
+import io.trino.connector.CatalogName;
+import io.trino.execution.scheduler.TestingNodeSelectorFactory.TestingNodeSupplier;
+import io.trino.memory.MemoryInfo;
+import io.trino.metadata.InternalNode;
+import io.trino.spi.HostAddress;
+import io.trino.spi.QueryId;
+import io.trino.spi.memory.MemoryPoolInfo;
+import io.trino.testing.assertions.Assert;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.Test;
+
+import java.net.URI;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static io.airlift.units.DataSize.Unit.GIGABYTE;
+import static io.trino.execution.scheduler.FallbackToFullNodePartitionMemoryEstimator.FULL_NODE_MEMORY;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertNotEquals;
+import static org.testng.Assert.assertTrue;
+
+// uses mutable state
+@Test(singleThreaded = true)
+public class TestFullNodeCapableNodeAllocator
```

Comment:
Should we have a test covering the logic around `sharedAllocatedMemory`?

Reply:
what exactly you think is not covered. The testcases which allocated "shared" note exercise exactly that.